components/ca-form.js

```diff
@@ -0,0 +1,1520 @@
+         * @param {Object} data, form data
+         * @returns {Boolean} if it's a valid object or not
+         */
+        isValid(data) {
+
+            // if validation has been disabled (for example, the Form Builder doesn't want/need it)
+            if (this.disableValidation === true) {
+                // then simply return true to indicate the form data is "valid"
+                return true;
+            }
+
+            // if a data object is not passed, get the current values
+            // eslint-disable-next-line no-param-reassign
+            data = data || this.getData();
+            const schema = this.selectedSchema;
+            let valid = true;
+
+            // eslint-disable-next-line no-restricted-syntax
+            for (const key in data) {
+
+                // ensure data and schema have matching key!
+                if (Object.prototype.hasOwnProperty.call(data, key) && Object.prototype.hasOwnProperty.call(schema.properties, key)) {
+
+                    // if the item has a regex patter, grab the raw string - we do this because parseInt strips zero's
+                    const inputEl = this.querySelector(`[name="${key}"]`);
+
+                    let value = (schema.properties[key].pattern) ? (inputEl && inputEl.value) || data[key] : data[key];
+
+                    value = (typeof value === 'string') ? value.trim() : value;
+
+                    const error = validateAgainstSchema(schema, key, value);
+
+                    if (error) {
+                        this.setError(key, error);
+                        valid = false;
+                    } else {
+                        this.clearError(key);
+                    }
+                }
+            }
+
+            // update session stored form data
+            if (this.persist && window.sessionStorage) {
+                sessionStorage[this.src] = JSON.stringify(this.getRawData());
+            }
+
+            return valid;
+        }
+
+        /**
+         * @description Saves the form data back to the server
+         * @returns {void} nothing
+         */
+        save() {
+
+            const createUrl = this.getAttribute('create-url') || '';
+            const updateUrl = this.getAttribute('update-url') || '';
+            const formData = this.getData();
+
+            this.clearValidationErrors();
+
+            // exit if not valid
+            if (!this.isValid(formData)) return false;
+
+            // execute update or create request
+            if (updateUrl !== '') {
+                componentSupport.request({
+                    url: updateUrl,
+                    type: 'PUT',
+                    dataType: 'json',
+                    data: formData,
+                    contentType: 'application/json'
+                })
+                    .then(data => {
+
+                        trace('Existing record created.');
```

Comment:
I agree it would be better to remove trace but I think we need to come up with a system for handling these messages and errors or remove them in the next stage. After all trace is just hiding console.log.